testcases/OpTestPCI.py

```diff
@@ -383,6 +383,131 @@ def runTest(self):
+
+            def print_details(self):
+                print "%s, capability=%s, secondary=%s" %(self.get_id(), self.capability, self.secondary)
+                print "capspeed=%s, capwidth=%s, staspeed=%s, stawidth=%s" % (self.capspeed, self.capwidth, self.staspeed, self.stawidth)
+
+            def get_id(self):
+                return "%s:%s:%s" % (self.domain, self.primary, self.slotfunc)
+
+        # Checking if two devices are linked together
+        def devicesLinked(upstream,downstream):
+            if upstream.domain == downstream.domain:
+                if upstream.secondary == downstream.primary:
+                    if upstream.capability == "Root":
+                        if downstream.capability == "Upstream":
+                            return True
+                        if downstream.capability == "Endpoint":
+                            return True
+                    if upstream.capability == "Downstream":
+                        if downstream.capability == "Endpoint":
+                            return True
+            return False
+
+        # Checking if LnkSta matches LnkCap
+        def optimalLink(upstream, downstream):
+            if upstream.capspeed > downstream.capspeed:
+                optimal_speed = downstream.capspeed
+            else:
+                optimal_speed = upstream.capspeed
+            if upstream.capwidth > downstream.capwidth:
+                optimal_width = downstream.capwidth
+            else:
+                optimal_width = upstream.capwidth
+            if optimal_speed > upstream.staspeed:
+                return False
+            if optimal_width > upstream.stawidth:
+                return False
+            return True
+
+        device_list = []
+
+        # Filling device objects' details
+        for device in device_ids:
+            device_info = self.cv_SYSTEM.console.run_command("lspci -s %s -vv" % (device))
+            device_list.append(Device(device_info))
+
+        checked_devices = []
+        suboptimal_links = []
+
+        # Searching through devices to check for links and testing to see if they're optimal
+        for device in device_list:
+            if device not in checked_devices:
+                checked_devices.append(device)
+                for endpoint in device_list:
+                    if endpoint not in checked_devices:
+                        if devicesLinked(device, endpoint):
+                            print "checking link between %s and %s" % (device.get_id(), endpoint.get_id())
+                            device.print_details()
+                            endpoint.print_details()
+                            print ""
+                            checked_devices.append(endpoint)
+                            if not optimalLink(device, endpoint):
+                                suboptimal_links.append((device, endpoint))
+
+        # asset suboptimal list is empty
+        self.assertEqual(len(suboptimal_links), 0)
```

Comment:
I ran this on my witherspoon and got this failure: 
```
======================================================================
FAIL [124.173s]: runTest (testcases.OpTestPCI.TestPciLink)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/stewart/op-test-framework/testcases/OpTestPCI.py", line 505, in runTest
    self.assertEqual(len(suboptimal_links), 0)
AssertionError: 2 != 0

----------------------------------------------------------------------
Ran 1 test in 124.174s
```

Before thinking of if it's a real failure or not, I'm tihnking we should be able to get a good error message out. If you add a `msg="Suboptimal links found: {}".format(str(suboptimal_links))` to the `assertEqual` call and do some __repr__ or __str__ magic methods for the Device class, you should be able to get something nice out as an error message when the test fails